### PR TITLE
Make guns server authoritative and fix various gun issues

### LINKED ISSUE #957: Refactor weapons script
# Preliminary information

## Category
TODO

# Report
## Current Behavior
Currently the weapons script doesn't handle melee weapons.  It also has hard coded input behavior.

## Expected/Wanted/Requested Behavior
The weapons script should be broken up.  
- [ ] It should become a Weapon base script, with two scripts inheriting from it:  Gun and MeleeWeapon.  
- [ ] Weapon should inherit from InputTrigger, ~~then tied into the input system as described in ticket #953 .~~ 
– _input system is yet to be redesigned_
- [ ] The melee weapon handling in HealthInput should be removed, and handled in the MeleeWeapon script.

### LINKED ISSUE #989: Revamp Firearms
# Preliminary information

## Category
TODO, related to #957 

# Report
## Current Behavior
Firearms are prone to speedhack and have client authority, fire rate and mag size can be tampered with easily

## Expected/Wanted/Requested Behavior
Firearms should have server authority!

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
1. Host in editor, join as build 
2. Get a P90 as build
3. Enable CE speedhack for build at 3x speed
4. Empty mag in a second, verify that it looked the same on server

### LINKED ISSUE #1123: Firearms: NRE when client's mag is ejected while shooting
# Preliminary information

## Category
BUG

## Bug: Release or Development?
Development

# Report
## Current Behavior
Sometimes when client sends shoot requests and mag is already ejected NREs appear.

## Expected/Wanted/Requested Behavior
No NREs

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
Kinda hard to replicate, might take several attempts
1. Host in editor, join as build (reproduced with 200 ping setting, syndicates)
2. Pick up machine gun as client and keep shooting till you run out of ammo, peeking at server console
3. Eventually these will apear:
`[12:42:21:761] NullReferenceException: Object reference not set to an instance of an object
Weapons.Weapon.Shoot (UnityEngine.GameObject shooter, UnityEngine.Vector2 direction, System.String bulletName, BodyPartType damageZone, System.Boolean isSuicideShot) (at Assets/Scripts/Weapons/Weapon.cs:383)
Weapons.Weapon.ServerShoot (UnityEngine.GameObject shotBy, UnityEngine.Vector2 direction, System.String bulletName, BodyPartType damageZone, System.Boolean isSuicideShot) (at Assets/Scripts/Weapons/Weapon.cs:366)
RequestShootMessage+<Process>c__Iterator0.MoveNext () (at Assets/Scripts/Messages/Client/RequestShootMessage.cs:32)
UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) (at C:/buildslave/unity/build/Runtime/Export/Coroutines.cs:17)
`

## OS used:
Windows

## Network
Localhost editor + 200ms build

### LINKED ISSUE #1399: Critical clients can still shoot
# Description
Clients who are critical and have a gun in their active hand can still shoot. Note that this can only be reproduced after #1376 is merged.

## Category
BUG

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
1. Host in editor, connect in client.
2. As client, grab a gun and shoot yourself until you go unconscious.
3. Try shooting while unconscious (listen for the shoot sfx)
suicide
.

Most of the logic related to this happens in PlayerNetworkActions.SetConsciousState. It looks like the host properly adjusts the player's state, but does not communicate that the player is unconscious to the client. Probably just need to add another rpc method which sends this info to the client and updates the client's consciousness state to unconscious.

### PR DESCRIPTION
**Pull request by chairbender**
### Purpose
Fixes #1123 
Fixes #989 
Fixes #957
Fixes #1399 

Makes guns server authoritative while still preserving client prediction, also fixes some various gun issues as a side effect.

### Open Questions and Pre-Merge TODOs

- [x]  This fix is tested on the same branch it is PR'ed to.
- [x]  I correctly commented my code
- [x]  My code is indented with tabs and not spaces
- [x]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [x]  This PR does not bring up any new compile errors
- [x]  This PR has been tested in editor
- [x]  This PR has been tested in multiplayer

### Notes:
Makes guns server authoritative while still preserving client prediction to make shooting feel responsive for the client. Accomplishes this mostly by putting gun-related actions (shooting, unloading, reloading) into a queue and being careful about when client and server ammo counts are synced.

Also ensures that accuracy deviation is consistent between the client and server (while still being server authoritative) so that client predicted shots deviate the same way as the server's version of the shot. Does this by using the magazine net ID as the seed for the RNG.

### CODE REVIEW

**Comment by chairbender** [comment]
just noticed a few remaining things to clean up, DNM for a sec

**Comment by chairbender** [comment]
Fixed that stuff, ready for review now.

**Comment by chairbender** [comment]
FYI there are some issues (including security issues) around what happens when player becomes unable to shoot due to being critical, planning on tackling that in #1399 .

**Comment by chairbender** [comment]
Since people are busy with other stuff and haven't looked at this yet, I'll go ahead and just include #1399 in this since I'm already about done fixing it. DNM until I'm done testing that

**Comment by chairbender** [comment]
Ready for review again. 1399 should be fixed as well in this PR

File: UnityProject/Assets/Scripts/Messages/Client/RequestShootMessage.cs [hunk 1]
```diff
@@ -3,43 +3,49 @@
...
 	}
 
 	public static RequestShootMessage Send(GameObject weapon, Vector2 direction, string bulletName,
 	                                       BodyPartType damageZone, bool isSuicideShot, GameObject shotBy)
 	{
 		RequestShootMessage msg = new RequestShootMessage {
-			Weapon = weapon ? weapon.GetComponent<NetworkIdentity>().netId : NetworkInstanceId.Invalid,
-			Direction = direction,
-			BulletName = bulletName,
+			Target = direction,
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/Messages/Client/RequestShootMessage.cs:46]: When I originally implemented shooting I created a ray cast on the client and captured the first point of collision and then sent the collision position to the server to determine direction there. This way it doesn't matter where the start point is (due to lag player could be further ahead on client then on the server) the bullet will always end up impacting the same spot. Lets see how this goes for a bit but in the future if there are too many inconsistencies then we should revert back to sending the end point of the trajectory and calculating direction via server.

File: UnityProject/Assets/Scripts/Weapons/MagazineBehaviour.cs [hunk 2]
```diff
@@ -1,23 +1,72 @@
...
+	(such as when reloading or when picking up a weapon with this mag in it).
+	*/
+	[SyncVar] private int serverAmmoRemains;
+	private int clientAmmoRemains;
+	/// <summary>
+	/// Remaining ammo in the clip. On the server, this will always be get/set using the server
+	/// authoritative value. On the client, this will use our local ammoRemains count, which can be synced
+	/// with the server-authoritative count using SyncClientAmmoRemainsWithServer()
+	/// </summary>
+	public int ammoRemains
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/Weapons/MagazineBehaviour.cs:25]: I like how you've done the server / client checks inside the property. very nice

**Review by DooblyNoobly — APPROVED** [review]

File: UnityProject/Assets/Scripts/Messages/Client/RequestShootMessage.cs (hunk 1 above)
Comment (DooblyNoobly) [reply UnityProject/Assets/Scripts/Messages/Client/RequestShootMessage.cs:46]: Actually seems to work pretty good with high lag. I tested at 800ms and didn't notice any inconsistencies with damage coming from server

**Pull request merged**